dali/util/uri.cc

```diff
@@ -0,0 +1,142 @@
+// Copyright (c) 2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/util/uri.h"
+#include <utility>
+#include <stdexcept>
+
+namespace dali {
+
+inline bool allowed_scheme_char(char c) {
+  return std::isalnum(c) || c == '.' || c == '+' || c == '-';
+}
+
+inline bool allowed_char(char c) {
+  // See https://en.wikipedia.org/wiki/Uniform_Resource_Identifier
+  // gen-delims: :, /, ?, #, [, ], and @
+  // sub-delims: !, $, &, ', (, ), *, +, ,, ;, and =
+  // unreserved characters (uppercase and lowercase letters, decimal digits, -, ., _, and ~)
+  // the character %
+  static const std::string gen_delims = ":/?#[]@";
+  static const std::string sub_delims = "!$&'()*+,;=";
+  static const std::string unreserved = "-._~";
+  return (std::isalnum(c)
+    || gen_delims.find(c) != std::string::npos
+    || sub_delims.find(c) != std::string::npos
+    || unreserved.find(c) != std::string::npos);
+}
+
+URI URI::Parse(std::string uri) {
+  // See https://en.wikipedia.org/wiki/Uniform_Resource_Identifier
+  URI parsed;
+  parsed.uri_ = std::move(uri);
+  parsed.valid_ = true;
+  size_t len = parsed.uri_.length();
+  const char* p_start = parsed.uri_.data();
+  const char* p_end = parsed.uri_.data() + len;
+  const char* p = p_start;
+
+  // Scheme
+  parsed.scheme_start_ = p - p_start;
+
+  if (!std::isalpha(*p)) {
+    parsed.valid_ = false;
+    parsed.err_msg_ = "First character should be a an alpha character";
+    return parsed;
+  }
+  while (*p != '\0' && *p !=  ':') {
+    if (!allowed_scheme_char(*p)) {
+      parsed.valid_ = false;
+      parsed.err_msg_ = "Invalid character found (" + std::to_string(*p) + ") in scheme";
```

Comment:
I don't think that `to_string` does anything here. You could just add a `char` to a `string`. Still, even better would be to add some kind of escaping function for special characters, like `\a`, `\t`, \n` and possibly everything non-ASCII.